File: android/chat-core/src/main/kotlin/jp/co/altive/chat/ChatLinkPreview.kt
package jp.co.altive.chat

import java.net.URI

/** 利用アプリの画像ローダーへそのまま渡すリンクプレビュー画像。 */
data class ChatLinkPreviewImage(
  /** AltiveChatが解釈しない画像resource。 */
  val resource: String,
  val pixelWidth: Int,
  val pixelHeight: Int,
) {
  /** 空resourceや壊れた寸法を画像loaderへ渡さないための表示可否。 */
  val isDisplayable: Boolean
    get() = resource.isNotBlank() && pixelWidth > 0 && pixelHeight > 0

  /** 正の縦横寸法を両方保持している場合だけ返す縦横比。 */
  val aspectRatio: Float?
    get() = if (isDisplayable) pixelWidth.toFloat() / pixelHeight else null
}

/** Webリンクプレビューへ表示する、利用アプリが検証済みの値。 */
data class ChatLinkPreview(
  val sourceUrl: String,
  val title: String,
  val description: String? = null,
  val siteName: String? = null,
  val image: ChatLinkPreviewImage? = null,
) {
  /** HTTP(S) URLと空でないtitleを持ち、安全に表示できるか。 */
  val isDisplayable: Boolean
    get() = title.isNotBlank() && title.length <= 200 &&
      (description == null || description.length <= 500) &&
      (siteName == null || siteName.length <= 100) &&
      normalizeChatLinkPreviewUrl(sourceUrl) != null
}

/** テキスト本文で最初に現れるWeb URLを選択し、scheme省略時はHTTPSへ正規化する。 */
object ChatLinkPreviewParser {
  private val webUrlPattern = Regex(
    "(?<!:)https?://[^\\s<>\\\"'、。，．！？；：]+|" +
      "(?<![@:/\\p{L}\\p{N}._-])(?:www\\.)?" +
      "(?:[\\p{L}\\p{N}](?:[\\p{L}\\p{N}-]{0,61}[\\p{L}\\p{N}])?\\.)+" +
      "[\\p{L}]{2,63}(?::\\d{2,5})?(?:/[^\\s<>\\\"'、。，．！？；：]*)?",
    RegexOption.IGNORE_CASE,
  )
  private val trailingPunctuation =
    setOf('.', ',', '!', '?', ';', ':', '、', '。', '，', '．', '！', '？', '；', '：', ')', ']', '}', '＞', '》', '」', '』', '】')

  /** 本文中の先頭Web URLをHTTP(S)へ正規化して返す。不正な候補しかない場合はnull。 */
  fun firstUrl(text: String): String? = webUrlPattern.findAll(text)
    .map { it.value.trimEnd { character -> character in trailingPunctuation } }
    .map { value ->
      val candidate = if (value.startsWith("http://", ignoreCase = true) ||
        value.startsWith("https://", ignoreCase = true)
      ) {
        value
      } else {
        "https://$value"
      }
      normalizeChatLinkPreviewUrl(candidate)
    }
    .filterNotNull()
    .firstOrNull()
}

/** resolverと保存値の比較に使うHTTP(S) URLへ正規化する。 */
fun normalizeChatLinkPreviewUrl(value: String): String? = runCatching {
  val uri = URI(value)
  val scheme = uri.scheme?.lowercase()
  val host = uri.host?.lowercase()
  if ((scheme != "http" && scheme != "https") || host.isNullOrBlank()) return null
  val port = uri.port.takeUnless {
    (scheme == "http" && it == 80) || (scheme == "https" && it == 443)
  } ?: -1
  URI(scheme, uri.userInfo, host, port, uri.path, uri.query, null).toASCIIString()
}.getOrNull()

File: lib/src/models/chat_link_preview.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/widgets.dart';

import 'sticker.dart';

/// Webリンクプレビューの画像参照。
@immutable
class ChatLinkPreviewImage extends Equatable {
  /// Webリンクプレビューの画像参照を作成する。
  const ChatLinkPreviewImage({
    required this.resource,
    required this.pixelWidth,
    required this.pixelHeight,
  }) : assert(pixelWidth > 0),
       assert(pixelHeight > 0);

  /// アプリ側の画像loaderへそのまま渡す不透明な参照値。
  final String resource;

  /// 画像のpixel幅。
  final int pixelWidth;

  /// 画像のpixel高さ。
  final int pixelHeight;

  /// 利用可能な縦横比。
  double get aspectRatio => pixelWidth / pixelHeight;

  /// 画像参照と寸法が表示可能かどうか。
  bool get isDisplayable =>
      resource.trim().isNotEmpty && pixelWidth > 0 && pixelHeight > 0;

  @override
  List<Object?> get props => [resource, pixelWidth, pixelHeight];
}

/// Webリンクプレビューの表示値。
@immutable
class ChatLinkPreview extends Equatable {
  /// Webリンクプレビューの表示値を作成する。
  const ChatLinkPreview({
    required this.sourceUrl,
    required this.title,
    this.description,
    this.siteName,
    this.image,
  });

  /// カードを操作した時に開くHTTP(S) URL。
  final Uri sourceUrl;

  /// 1〜200文字のタイトル。
  final String title;

  /// 最大500文字の説明。
  final String? description;

  /// 最大100文字のサイト名。
  final String? siteName;

  /// 任意の代表画像。
  final ChatLinkPreviewImage? image;

  /// 表示に必要な値が有効かどうか。
  bool get isDisplayable =>
      (sourceUrl.scheme == 'http' || sourceUrl.scheme == 'https') &&
      sourceUrl.host.isNotEmpty &&
      title.trim().isNotEmpty &&
      title.characters.length <= 200 &&
      (description == null || description!.characters.length <= 500) &&
      (siteName == null || siteName!.characters.length <= 100);

  @override
  List<Object?> get props => [sourceUrl, title, description, siteName, image];
}

/// 入力中の先頭Web URLを解決する処理。
typedef ChatLinkPreviewResolver =
    Future<ChatLinkPreview?> Function(Uri sourceUrl);

/// Webリンクプレビュー画像を構築する処理。
typedef ChatLinkPreviewImageBuilder =
    Widget Function(BuildContext context, ChatLinkPreviewImage image);

/// Webリンクを操作した時の処理。
typedef ChatWebLinkTapCallback = void Function(Uri sourceUrl);

/// Composerから送信する型付きの値。
@immutable
class ChatComposerSubmission extends Equatable {
  /// Composerから送信する型付きの値を作成する。
  const ChatComposerSubmission({
    required this.text,
    this.sticker,
    this.linkPreview,
  });

  /// 正規化済みの本文。
  final String text;

  /// 選択中のステッカー。
  final Sticker? sticker;

  /// 入力中に解決できた任意のリンクプレビュー。
  ///
  /// 楽観表示用であり、信頼済みの永続化値として扱わない。
  final ChatLinkPreview? linkPreview;

  @override
  List<Object?> get props => [text, sticker, linkPreview];
}
